More constructor caching and minor random optimisations

reaperrr opened the pull request:
Moved many trait look-ups from Tick (or other actions) to constructors.
Performance gains are likely negligible, but this also makes the affected code more consistent with traits and activities where this had already been done.
Additionally extracted `FlyOffMap` and `HarvestResource` activities into their own files, and moved harvesting-related activities into their own subfolder.

A couple of notes:
- There are a few more cases (aircraft attack activities and `Repariable` trait, for example) which I intentionally omitted since they're already part of #7324.
- I also omitted `FindResources` for now, because I had trouble dealing with that 'double'-constructor, my attempts compiled fine but crashed the game at launch. Feel free to make suggestions on how to deal with those constructors.

reaperrr committed:
Cache trait look-ups in the constructor where possible for Air activities/traits

reaperrr committed:
Misc constructor caching

Cache trait look-ups in constructor for various other traits and
activities.

RoosterDragon commented on OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs:
```diff
@@ -0,0 +1,40 @@
+			return this;
+		}
+
+		public override void Cancel(Actor self)
```
I realise this was just you moving a file, but an override that just calls the base method is a bit silly.

reaperrr replied:
So if I understand correctly, I can just remove the whole override?

penev92 replied:
Yes.

RoosterDragon commented on OpenRA.Mods.Common/Activities/CaptureActor.cs:
```diff
@@ -16,15 +16,19 @@ namespace OpenRA.Mods.Common.Activities
 	public class CaptureActor : Enter
 	{
 		readonly Actor actor;
+		readonly Building b;
```
Could we give this a better name since we're here?

RoosterDragon commented on OpenRA.Mods.Common/Activities/Sell.cs:
```diff
@@ -17,12 +17,19 @@ namespace OpenRA.Mods.Common.Activities
-		public override Activity Tick(Actor self)
+		readonly Health h;
+		readonly SellableInfo si;
+		readonly PlayerResources pr;
```
Could we give these better names since we're here?

RoosterDragon commented on OpenRA.Mods.Common/Traits/Transforms.cs:
```diff
@@ -46,14 +46,18 @@ public class Transforms : IIssueOrder, IResolveOrder, IOrderVoice
 		readonly Actor self;
 		readonly TransformsInfo info;
+		readonly Building b;
 		readonly BuildingInfo bi;
```
Could we get decent names for these as well.

reaperrr commented:
Updated and rebased.

I've removed harvester-related constructor caching from this PR to avoid rebase conflicts with #7658 (which is far more important). I have a local backup, I'll take another look once #7658 has been merged.

Unit158 commented on OpenRA.Mods.Cnc/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -67,7 +67,7 @@ public override bool Produce(Actor self, ActorInfo producee, string raceVariant)
 
 				a.QueueActivity(new Fly(a, Target.FromCell(w, self.Location + new CVec(9, 0))));
-				a.QueueActivity(new Land(Target.FromActor(self)));
+				a.QueueActivity(new Land(a, Target.FromActor(self)));
```
Can you change the variable name "a" to actor please.

RoosterDragon commented:
:+1:

obrakmann commented:
Code looks good, no issues in-game either, :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/8ad2d9ad0f0f7936322422a0761ba466743aa3ce%5E...8ad2d9ad0f0f7936322422a0761ba466743aa3ce)

pchote commented:
This has introduced a bunch of implicit yaml-ordering issues.  You can only safely cache traits from a trait ctor if it is explicitly `Requires<>`ed on the TraitInfo.
